### DISCUSSION TITLE
Can I run embedded wasmer on a wasm client?

### DISCUSSION THREAD

**Discussion by seniorjoinu** [wasmerio/wasmer#2617] [Q&A]
### Summary

My environment is webassembly itself (wasmtime). I write programs in rust, compile them to `wasm32-unknown-unknown` and then execute them. This flow works fine.
But now I need to write a program which can take another wasm program as an input, execute it and process the result. Such a wasmception!

How could I achieve that?

### Additional details

I tried to compile [hello_world](https://github.com/wasmerio/wasmer/blob/master/examples/hello_world.rs) to `wasm32-unknown-unknown`, but it looks like it can't be compiled that way.

**Comment by aristotaloss** [comment]
Seconding this, any way to get wasmer working as embedded library when targeting `wasm32-unknown-unknown` or even `wasm32-wasi`?

**Comment by JoshFourie** [comment]
**tldr**: I think that this is possible but painful and requires an overhaul of the `wasmer` design.

If you are working with `rust`, you can do this with `wasm-bindgen`  ([example](https://rustwasm.github.io/wasm-bindgen/examples/wasm-in-wasm.html)). I found managing imports/exports to `wasm` binaries painful and I like the `wasmer` api, so I would like support for this feature.

---
I saw in-tree that [wasmer/api/src/js](https://github.com/wasmerio/wasmer/tree/b30284897e049bb84ba8f01db0643daec62f57b6/lib/api/src/js) was recently added so that `wasmer` [will be compiled to WebAssembly to run in a JavaScript host](https://github.com/wasmerio/wasmer/blob/5e60296658a28e2f21df876688ea2af06346799c/lib/api/src/lib.rs#L286). I pulled that out of the `wasmer` tree and rolled a [patch](https://github.com/JoshFourie/wasmer-patch/tree/888455a4a64ff555200bca6c09afbe17b03e702a) to expose the `js` api ahead of the `wasmer` release schedule. 

This only worked for `wasm` binaries that were less than `4kb` and a [StackOverflow question](https://stackoverflow.com/a/45285081) recommended using either a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) or [WebAssembly.InstantiateStreaming](https://github.com/WebAssembly/design/blob/main/JS.md#webassemblyinstantiate). 

I did not want to use a web worker at this stage, so I had to work around a few things:

1. `WebAssembly.InstantiateStreaming` takes a `Promise` of a `Response` of the source of  `WebAssembly.Module` not a compiled `WebAssembly.Module`: [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/instantiateStreaming)
2. `wasmer::Instance` [constructor](https://github.com/wasmerio/wasmer/blob/5e60296658a28e2f21df876688ea2af06346799c/lib/api/src/js/instance.rs#L95) takes a `wasmer::Module` which [polyfills](https://github.com/wasmerio/wasmer/blob/5e60296658a28e2f21df876688ea2af06346799c/lib/api/src/js/module.rs#L171) the `ModuleTypeHints` struct with import/export information by [parsing the binary](https://github.com/wasmerio/wasmer/blob/5e60296658a28e2f21df876688ea2af06346799c/lib/api/src/js/module_info_polyfill.rs#L253), which is not available with `WebAssembly.InstantiateStreaming`

To get this to work, I had to build the `ModuleTypeHints` information without parsing the binary. In my case, this information was known at compile-time, so I rolled a new [resolver implementation](https://github.com/JoshFourie/wasmer-patch/blob/8806eedb7e8600eeee41318ae7d6a169383d8e41/src/js/import_object.rs#L185) and `wasmer::Instance` [constructor](https://github.com/JoshFourie/wasmer-patch/blob/8806eedb7e8600eeee41318ae7d6a169383d8e41/src/js/instance.rs#L108). See an [example](https://github.com/JoshFourie/wasm-in-wasm-experiment/blob/7f70cf52ba45cd814e4abaf2f2eefdb6a32a866f/runtime/src/lib.rs#L62).

I hacked out a [patch](https://github.com/JoshFourie/wasmer-patch) and an [example](https://github.com/JoshFourie/wasm-in-wasm-experiment) repository but I am not maintaining or working with that library. It does not properly handle results returned from `wasm` exports. I think that the `wasmer` team can probably fix this much more intelligently.
